src/kink.md

```diff
@@ -0,0 +1,548 @@
+```k
+requires "ekore.k"
+requires "file-util.k"
+requires "parser-util.k"
+```
+
+Syntax
+======
+
+We bypass K5 parsing to use k-lights. This is because K-light allows returning
+bubbles.
+
+```k
+module KINK-SYNTAX
+  syntax Any ::= r"([\\n\\r]|.)*" [token]
+endmodule
+```
+
+Configuration & Main Module
+===========================
+
+The K in K configuration has a "k" cell containing a definition, and a
+"pipeline" cell containing operations that map over the definition in the K
+cell. When an operation is at the top of the `<k>` cell, it must
+transform the declaration as needed.
+
+```k
+module KINK-CONFIGURATION
+  imports EKORE-ABSTRACT
+  imports SET
+  imports STRING-SYNTAX
+
+  syntax Any
+  syntax Declaration ::= "nullDecl"
+  syntax DeclCellSet
+  syntax DeclarationsCellFragment
+  configuration <pipeline> $PIPELINE:K ~> .K </pipeline>
+                <defn> $PGM:Any ~> .K </defn>
+                <definition>
+                   <defnAttrs format="[ %2 ]%n"> .Patterns </defnAttrs>
+                   <modules format="%2%n">
+                     <mod format="module %2%i%n%4%n%5%n%6%n%7%d%n %i%dendmodule %3%n%n"
+                          multiplicity="*" type="Set">
+                       <name format="%2"> #token("UNNAMED", "ModuleName"):ModuleName </name>
+                       <attributes format="[ %2 ]"> .Patterns </attributes>
+                       <imp format="%2"> .Declarations </imp>
+                       <declarations format="%2">
+                         <decl format="%2%n" multiplicity="*" type="Set"> nullDecl </decl>
+                       </declarations>
+                       <k> .K </k>
```

Comment:
That's interesting. I wasn't expecting to have a k cell in every module.
I can see how this can work for what we have currently but somehow feels off.
Can you explain why you chose this solution?

Reply:
It allows pipeline steps to be executed in parallel for each module. There is nothing unusual about this -- its a common approach for implementing threads. We will however need to be careful about checking for dependencies between passes. e.g. we should not try to generate grammars until after `#flattenProductions` has happened in every imported module.